README.md
Homie Device
============

[![NPM](https://nodei.co/npm/homie-device.svg?downloads=true&downloadRank=true)](https://nodei.co/npm/homie-device/)&nbsp;&nbsp;
[![Build Status](https://secure.travis-ci.org/microclimates/homie-device.svg?branch=master)](https://travis-ci.org/microclimates/homie-device)&nbsp;&nbsp;
[release notes](https://github.com/microclimates/homie-device/blob/master/History.md)

This is a NodeJS port of the [Homie convention](https://github.com/marvinroger/homie) for lightweight IoT device interaction on an [MQTT](https://en.wikipedia.org/wiki/MQTT) message bus.

It is modeled after the great work done for [ESP8266 devices](https://github.com/marvinroger/homie-esp8266), with the goal of not only following the Homie Convention, but at adhering to the implementation interface so both message and API interfaces are familiar.

It's great for mixing ESP8266 devices with Raspberry Pi and other linux or windows based systems on the same MQTT network.

Features
--------

* Device, Node, and Property with ESP8266-like interface
* Auto MQTT connect with optional username/password
* Auto MQTT re-connect
* Device config matching ESP8266 config JSON
* Periodic $stats/uptime publishing
* $online will
* Device topic events
* Broadcast message events
* Periodic stats interval events
* Device/node/property announcement on connect
* Property send with retained value
* Settable properties
* Property ranges
* Lightweight
* Full test coverage

Quick Start
-----------

First, get a local [MQTT broker](https://mosquitto.org/download/) running and a window opened subscribing to the 'devices/#' topic.

Next, add these lines to an index.js file and run it:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('bare-minimum');
myDevice.setup();
```

Congratulations, you've got a running Homie device! 

Take a look at the messages on your MQTT bus under devices/bare-minimum. Then ctrl-c the program and watch the broker administer the will.

Setting Firmware
----------------

To publish the firmware name & version, call `myDevice.setFirmware()`:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('bare-minimum');
myDevice.setFirmware('nodejs-test', '0.0.1');
myDevice.setup();
```

Adding a Node
-------------

Devices aren't much use until they have some nodes and properties. Place the following into the index.js file and run it again:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');
var myNode = myDevice.node('my-node', 'test node friendly name', 'test-node');
myNode.advertise('my-property-1').setName('Friendly Prop Name').setUnit('W').setDatatype('integer');
myDevice.setup();
```

Publishing a Property
---------------------

Publishing properties has the same interface as the Homie ESP8266 implementation:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');
var myNode = myDevice.node('my-node', 'test node friendly name', 'test-node');
myNode.advertise('my-property-1').setName('Friendly Prop Name').setUnit('W').setDatatype('integer');
myDevice.setup();

myNode.setProperty('my-property-1').send('property-value');
```

Settable Properties
-------------------

To set properties from MQTT messages, add a setter function when advertising the property:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');
var myNode = myDevice.node('my-node', 'test node friendly name', 'test-node');
myNode.advertise('my-property-1').setName('Friendly Prop Name').setUnit('W').setDatatype('string').settable(function(range, value) {
  myNode.setProperty('my-property-1').setRetained().send(value);
});
myDevice.setup();
```

Once running, publish a message to the `devices/my-device/my-node/my-property-1/set` topic.

Array Nodes
----------------

Array nodes are also supported by passing lower and upper bounds on the range when creating the node.

```
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');

var lowerBound = 0;
var upperBound = 10;
var myNode = myDevice.node('my-node', 'test node friendly name', 'test-node', lowerBound, upperBound);
```

Publishing a property to a specific node index can be done like this:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');
var myNode = myDevice.node('my-node', 'test node friendly name', 'test-node', 0, 10);
myNode.advertise('my-property-1');
myDevice.setup();

// Publishes 'property-value' to the 'devices/my-device/my-node_2/my-property-1' topic.
myNode.setProperty('my-property-1').setRange(2).send('property-value');
```

Setting a property on a specific node index can be done like this:

```
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');
var myNode = myDevice.node('my-node', 'test node friendly name', 'test-node', 0, 10);
myNode.advertise('my-property-2').settable(function(range, value) {
  var index = range.index;
  myNode.setProperty('my-property-2').setRange(index).send(value);
});
myDevice.setup();
```

Now publish a message to the `devices/my-device/my-node_8/my-property-2/set` topic.

Device Messages
---------------

Incoming messages to the device emit `message` events. You can listen for all messages to the `devices/my-device/#` topic like this:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');

myDevice.on('message', function(topic, value) {
  console.log('A message arrived on topic: ' + topic + ' with value: ' + value);
});

myDevice.setup();
```

Now publish a message to the `devices/my-device/my-node/my-property-2_8/set` topic.

Topic Messages
--------------

You can listen for specific incoming topics by adding a listener to the `message:{topic}` event for the device:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');

myDevice.on('message:my/topic', function(value) {
  console.log('A message arrived on the my/topic topic with value: ' + value);
});

myDevice.setup();
```

Now publish a message to the `devices/my-device/my/topic` topic.

Broadcast Messages
------------------

You can listen to all `devices/$broadcast/#` messages by adding a listener to the `broadcast` event:

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('my-device');

myDevice.on('broadcast', function(topic, value) {
  console.log('A broadcast message arrived on topic: ' + topic + ' with value: ' + value);
});

myDevice.setup();
```

Now publish a broadcast message to the `devices/$broadcast/some-topic` topic. All homie devices are exposed to broadcast messages.

Device Settings
---------------

To configure your device with external settings, pass a full or partial config object to the HomieDevice constructor. If you've worked with the Homie ESP8266 implementation, this will be familiar:

```javascript
var HomieDevice = require('homie-device');
var config = {
  "name": "Bare Minimum",
  "device_id": "bare-minimum",
  "mqtt": {
    "host": "localhost",
    "port": 1883,
    "base_topic": "devices/",
    "auth": false,
    "username": "user",
    "password": "pass"
  },
  "settings": {
    "percentage": 55
  }
}
var myDevice = new HomieDevice(config);
myDevice.setup();
```

Connection
----------

The Homie device maintains an `isConnected` (true/false) state, and emits 
a `connect` and `disconnect` event as the device becomes connected with MQTT.

Quiet Setup
-----------

If you don't want the startup message, pass the quiet flag of `true` to setup `myDevice.setup(true)`.

```javascript
var HomieDevice = require('homie-device');
var myDevice = new HomieDevice('bare-minimum');
myDevice.setup(true);
```

Contributors
------------
<table id="contributors"><tr>
<td><img width="124" src="https://avatars2.githubusercontent.com/u/373538?v=4"><br/><a href="https://github.com/lorenwest">lorenwest</a></td><td><img width="124" src="https://avatars0.githubusercontent.com/u/7427179?v=4"><br/><a href="https://github.com/marcus-garvey">marcus-garvey</a></td><td><img width="124" src="https://avatars0.githubusercontent.com/u/7627635?v=4"><br/><a href="https://github.com/rozpuszczalny">rozpuszczalny</a></td><td><img width="124" src="https://avatars1.githubusercontent.com/u/5964259?s=460&v=4"><br/><a href="https://github.com/carterzenk">carterzenk</a></td></tr></table>

License
-------

May be freely distributed under the [MIT license](https://raw.githubusercontent.com/microclimates/homie-device/master/LICENSE).

Copyright (c) 2017-2018 Loren West 
[and other contributors](https://github.com/microclimates/homie-device/graphs/contributors)



History.md
3.0.1 / 2019-05-08
==================

  * publish stats interval at each interval - @carterzenk
  * remove optional comments from setters that are not optional - @carterzenk
  * Updated readme to show examples of working with array nodes - @carterzenk
  * Added additional test coverage for 3.0.1 features - @carterzenk
  * Published to NPM as 3.0.1 - @lorenwest


0.4.0 / 2019-01-08
==================

  * Publish the node ID list - @jpg0
  * Differentiate node name from node ID - @rozpuszczalny

0.3.2 / 2018-01-03
==================

  * Added travis_retry because sometimes Travis hangs

0.3.1 / 2018-01-03
==================

  * Updating Travis node versions to keep Travis happy

0.3.0 / 2018-01-03
==================

  * Added isConnected internal state to HomieNode

0.2.1 / 2017-09-07
==================

  * Fixed range property topic name

0.1.3 / 2017-09-01
==================

  * Changed event order to node connects first, device connect after

0.1.2 / 2017-09-01
==================

  * Rounded out implementation
  * Added range properties
  * Added message events
  * Added broadcast messages
  * Full test suite
  * Better documentation

0.1.1 / 2017-09-31
==================

  * Minimum implementation
  * No documentation


0.1.0 / 2017-08-30
==================

  * Initial publish


test/mqttStub.js
var _ = require('lodash');
var EventEmitter = require('events').EventEmitter;

var MQTTClientStub = module.exports = function(serverUrl, opts) {
  var t = this;
  t.serverUrl = serverUrl;
  t.opts = opts;
  t.publishedMsgs = [];
}
require('util').inherits(MQTTClientStub, EventEmitter);
var proto = MQTTClientStub.prototype;

MQTTClientStub.connect = function(serverUrl, opts) {
  var client = new MQTTClientStub(serverUrl, opts);
  setTimeout(function(){
    client.emit('connect');
  },1);
  return client;
}

proto.publish = function(topic, msg, opts) {
  var t = this;
  t.publishedMsgs.push({topic:topic, msg:msg, opts:opts});
  setTimeout(function(){
    // Auto subscribe to all published messages
    t.emit('message', topic, msg);
  },1);
}

proto.end = function() {
  var t = this;
  t.emit('close');
}

proto.getPublishedMsg = function(topic) {
  var t = this;
  return _.filter(t.publishedMsgs, {topic: topic});
}

proto.subscribe = function(topic) {
  var t = this;
}


test/2-HomieNode.js
var expect = require("chai").expect;
var proxyquire = require('proxyquire').noPreserveCache();
var quietSetup = true;
var mqtt = require('./mqttStub');
// var mqtt = require('mqtt'); // Uncomment to test with a live broker
var HomieDevice = proxyquire('..', {mqtt: mqtt});
var HomieNode = require('../lib/homieNode');

describe("Homie Node", function() {

  describe("Instantiation", function() {

    var testDevice = new HomieDevice('homie-device-test');
    var testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node');

    it("creates a class of type HomieNode", function() {
      expect(testNode1).to.be.an.instanceOf(HomieNode);
    });
    it("constructor arguments are correct", function() {
      expect(testNode1.name).to.equal('test-node-1');
      expect(testNode1.type).to.equal('test-node');
    });

  });

  describe("MQTT Connection", function() {

    var testDevice = new HomieDevice('homie-device-test');
    var testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node');

    it("emits the connect message on setup()", function(done) {
      testNode1.on('connect', function() {
        done();
      });
      testDevice.setup(quietSetup);
    });

    it("emits the disconnect message on end()", function(done) {
      testNode1.on('disconnect', function() {
        done();
      });
      testDevice.end();
    });

    it("publishes the node on connect", function(done) {
      testDevice = new HomieDevice('homie-device-test');
      testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node');
      testDevice.on('message:test-node-1/$type', function(msg) {
        expect(msg).to.equal('test-node');
        testDevice.end();
        done();
      });
      testDevice.setup(quietSetup);
    });

    it("publishes the node name on connect", function(done) {
      testDevice = new HomieDevice('homie-device-test');
      testNode1 = testDevice.node('test-node-1', 'friendly Name', 'test-node');
      testDevice.on('message:test-node-1/$name', function(msg) {
        expect(msg).to.equal('friendly Name');
        testDevice.end();
        done();
      });
      testDevice.setup(quietSetup);
    });

    it("publishes the node list on connect", function(done) {
      testDevice = new HomieDevice('homie-device-test');
      testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node');
      testDevice.on('message:$nodes', function(msg) {
        expect(msg).to.equal('test-node-1');
        testDevice.end();
        done();
      });
      testDevice.setup(quietSetup);
    });

    it("publishes the node list with array syntax on connect", function(done) {
      testDevice = new HomieDevice('homie-device-test');
      testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node',0,1);
      testDevice.on('message:$nodes', function(msg) {
        expect(msg).to.equal('test-node-1[]');
        testDevice.end();
        done();
      });
      testDevice.setup(quietSetup);
    });

  });

  describe("Stats publishing", function() {

    var testDevice = new HomieDevice('homie-device-test');
    var testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node');
    var numMsgs = 0;
    it("publishes node stats at a determined interval", function(done) {
      testNode1.on('stats-interval', function() {
         if (++numMsgs == 3) {
           testDevice.end();
           done();
         }
      });
      testDevice.statsInterval = .1;
      testDevice.setup(quietSetup);
    });

  });

});


test/homieDevice.js
var expect = require("chai").expect;
var proxyquire = require('proxyquire');
var mqtt = require('./mqttStub');
// var mqtt = require('mqtt'); // Uncomment to test with a live broker
var HomieDevice = proxyquire('..', {mqtt: mqtt});

describe("Homie Device", function() {

  describe("Instantiation", function() {

    var testDevice = new HomieDevice('bare-minimum');
    it("creates a class of the correct type", function() {
      expect(testDevice).to.be.an.instanceOf(HomieDevice);
    });
    it("creates a class with a string device_id constructor", function() {
      expect(testDevice.config.device_id).to.equal('bare-minimum');
    });

    var testDevice = new HomieDevice({name: 'Bare Minimum', device_id: 'bare-minimum', settings:{arg1:'value1'}});
    it("creates a class with a config object constructor", function() {
      expect(testDevice.config.device_id).to.equal('bare-minimum');
    });

    it("allows the config object to maintain defaults", function() {
      expect(testDevice.config.mqtt.host).to.equal('localhost');
    });

    it("allows the config object to provide device specific settings", function() {
      expect(testDevice.config.settings.arg1).to.equal('value1');
    });

    it("allows the client to set firmware", function() {
      testDevice.setFirmware('my-firmware','0.2.1');
      expect(testDevice.firmwareName).to.equal('my-firmware');
      expect(testDevice.firmwareVersion).to.equal('0.2.1');
    });

  });

  describe("MQTT Connection", function() {

    var testDevice = new HomieDevice('bare-minimum');
    it("emits the connect message on setup()", function(done) {
      testDevice.on('connect', function() {
        done();
      });
      testDevice.setup();
    });
    it("emits the disconnect message on end()", function(done) {
      testDevice.on('disconnect', function() {
        done();
      });
      testDevice.end();
    });

  });

});


lib/homieNode.js
var _ = require('lodash');
var HomieProperty = require('./homieProperty');
var EventEmitter = require('events').EventEmitter;

var HomieNode = module.exports = function(homieDevice, name, friendlyName, type) {

  var t = this;
  t.props = {};
  t.name = name;
  t.type = type;
  t.friendlyName = friendlyName;
  t.homieDevice = homieDevice;
  t.mqttTopic = t.homieDevice.mqttTopic + '/' + t.name;

}

require('util').inherits(HomieNode, EventEmitter);
var proto = HomieNode.prototype;

proto.advertise = function(propName) {
  var t = this;
  return t.props[propName] = new HomieProperty(t, propName);
}

proto.advertiseRange = function(propName, start, end) {
  var t = this;
  var prop = t.props[propName] = new HomieProperty(t, propName);
  prop.setRange(start, end);
  return prop;
}

// Called on mqtt client connect
proto.onConnect = function() {
  var t = this;
  var mqttClient = t.homieDevice.mqttClient;

  // Announce properties to MQTT
  mqttClient.publish(t.mqttTopic + '/$type', t.type, {retain:true});
  mqttClient.publish(t.mqttTopic + '/$name', t.friendlyName, {retain:true});

  var ads = [];
  _.each(t.props, function(prop){
    ads.push(prop.name);
  })
  mqttClient.publish(t.mqttTopic + '/$properties', ads.join(','), {retain:true});

  // TODO add array handling
  // mqttClient.publish(t.mqttTopic + '/$array', '', {retain:true});

  _.each(t.props, function(prop){
    prop.onConnect();
  })

  t.emit('connect');
}

// Called on mqtt client disconnect
proto.onDisconnect = function() {
  var t = this;
  t.emit('disconnect');
}

// Called on every stats interval
proto.onStatsInterval = function() {
  var t = this;
  t.emit('stats-interval');
}

// This name isn't very good (should be getProperty), but it matches the esp8266 homie implementation
proto.setProperty = proto.getProperty = function(propName) {
  var t = this;
  return t.props[propName];
}


lib/homieProperty.js
var _ = require('lodash');
var HomieProperty = module.exports = function(homieNode, name) {
  var t = this;
  t.name = name;
  t.friendlyName = null;
  t.setter = null;
  t.isSubscribedToSet = false;
  t.rangeStart = null;
  t.rangeEnd = null;
  t.retained = true;
  t.rangeIndex = null;
  t.homieNode = homieNode;
  t.mqttTopicProperty = t.homieNode.mqttTopic + '/' + t.name;
  t.format = null;
  t.datatype = null;
  t.unit = null;
}
var proto = HomieProperty.prototype;

proto.onConnect = function() {
  var t = this;
  var mqttClient = t.homieNode.homieDevice.mqttClient;

  mqttClient.publish(t.mqttTopicProperty + '/$name', t.friendlyName, {retain:true});
  mqttClient.publish(t.mqttTopicProperty + '/$retained',  t.retained ? "true" : "false", {retain:true});
  mqttClient.publish(t.mqttTopicProperty + '/$settable',  t.setter ? "true" : "false", {retain:true});

  if(t.unit !== null) {
    mqttClient.publish(t.mqttTopicProperty + '/$unit', t.unit, {retain:true});
  }

  if(t.datatype !== null) {
    mqttClient.publish(t.mqttTopicProperty + '/$datatype', t.datatype, {retain:true});
  }

  if(t.format !== null) {
    mqttClient.publish(t.mqttTopicProperty + '/$format', t.format, {retain:true});
  }
  
}

proto.setName = function(friendlyName) {
  var t = this;
  t.friendlyName = friendlyName;
  return t;
}

proto.setUnit = function(unit) {
  var t = this;
  t.unit = unit;
  return t;
}

proto.setDatatype = function(datatype) {
  var t = this;
  t.datatype = datatype;
  return t;
}

proto.setFormat = function(format) {
  var t = this;
  t.format = format;
  return t;
}

proto.settable = function(setter) {
  var t = this;
  t.setter = setter;
  return t;
}

proto.setRetained = function(val) {
  var t = this;
  t.retained = val;
  return t;
}

proto.send = function(val) {
  var t = this;
  var mqttClient = t.homieNode.homieDevice.mqttClient;
  var topic = t.mqttTopicProperty;
  if (t.rangeIndex !== null) {
    topic += '_' + t.rangeIndex;
  }
  mqttClient.publish(topic, val, {retain: t.retained});
  t.retained = false;
  t.rangeIndex = null;
  return t;
}


lib/homieDevice.js
var _ = require('lodash');
var pkgJson = require('../package.json');
var homieVersion = '3.0.1';
var homieImplName = 'nodejs:' + pkgJson.name;
var homieImplVersion = pkgJson.version;
var mqtt = require('mqtt');
var HomieNode = require('./homieNode');
var EventEmitter = require('events').EventEmitter;

const DEFAULT_CONFIG = {
  "name": "",
  "device_id": "unknown",
  "mqtt": {
    "host": "localhost",
    "port": 1883,
    "base_topic": "homie/",
    "auth": false,
    "username": null,
    "password": null,
  },
  "settings": {
  },
  "ip": null,
  "mac": null
}

/* Constructor
 *
 * Construct a new HomieDevice with a device_id or a config object containing some or all of:
 *
 * {
 *   "name": "Bare Minimum",
 *   "device_id": "bare-minimum",
 *   "mqtt": {
 *     "host": "localhost",
 *     "port": 1883,
 *     "base_topic": "devices/",
 *     "auth": false,
 *     "username": "user",
 *     "password": "pass"
 *   },
 *   "settings": {
 *     "percentage": 55 // device settings
 *   }
 * }
 *
 * Would like, but not implemented:
 *
 *   "wifi": {
 *     "ssid": "ssid",
 *     "password": "pass"
 *   },
 *   "ota": {
 *     "enabled": true
 *   }
 *
 */
var HomieDevice = module.exports = function(config) {

  var t = this;
  if (_.isString(config)) {
    config = {name: config, device_id: config};
  }
  t.config = _.extend({}, DEFAULT_CONFIG, config);

  t.name = t.config.device_id;
  t.mqttTopic = t.config.mqtt.base_topic + t.config.device_id;
  t.startTime = Date.now();

  t.nodes = {};
  t.firmwareName = null;
  t.firmwareVersion = null;
  t.statsInterval = 60;
  t.isConnected = false;
  t.mqttClient = null;
  t.friendlyName = t.config.name;
}

require('util').inherits(HomieDevice, EventEmitter);
var proto = HomieDevice.prototype;

proto.setFirmware = function(firmwareName, firmwareVersion) {
  var t = this;
  t.firmwareName = firmwareName;
  t.firmwareVersion = firmwareVersion;
}

proto.node = function(name, friendlyName, type, startRange, endRange) {
  var t = this;
  return t.nodes[name] = new HomieNode(t, name, friendlyName, type, startRange, endRange);
}

// Start the device
proto.setup = function(quiet) {
  var t = this;
  var mqttServer = 'mqtt://' + t.config.mqtt.host + ':' + t.config.mqtt.port;
  var opts = {
    will: {
      topic: t.mqttTopic + '/$state',
      payload: 'lost',
      qos: 0,
      retain: true
    }
  }
  if (t.config.mqtt.auth) {
    opts.username = t.config.mqtt.username;
    opts.password = t.config.mqtt.password;
  }
  t.mqttClient = mqtt.connect(mqttServer, opts);

  t.mqttClient.on('connect', function(connack) {
    t.onConnect();
  })

  t.mqttClient.on('close', function() {
    t.onDisconnect();
  })

  t.mqttClient.on('message', function (topic, message) {
    if(message != null) {
      t.onMessage(topic, message.toString());
    } else {
      t.onMessage(topic, null);
    }
    
  })

  t.mqttClient.subscribe(t.mqttTopic + '/#');
  t.mqttClient.subscribe(t.config.mqtt.base_topic + '$broadcast/#');

  if (!quiet) {
    console.log('Connected Homie ' + t.mqttTopic + ' to ' + mqttServer);
  }
}

// Stop the device
proto.end = function() {
  var t = this;
  t.mqttClient.publish(t.mqttTopic + '/$state', 'disconnected');
  t.mqttClient.end();
}

// Called on mqtt client connect
proto.onConnect = function() {
  var t = this;
  t.isConnected = true;

  // Advertise device properties
  t.mqttClient.publish(t.mqttTopic + '/$state', 'init', {retain:true});
  t.mqttClient.publish(t.mqttTopic + '/$homie', homieVersion, {retain:true});
  t.mqttClient.publish(t.mqttTopic + '/$implementation', homieImplName, {retain:true});
  t.mqttClient.publish(t.mqttTopic + '/$implementation/version', homieImplVersion, {retain:true});
  if (t.firmwareName) {
    t.mqttClient.publish(t.mqttTopic + '/$fw/name', t.firmwareName, {retain:true});
    t.mqttClient.publish(t.mqttTopic + '/$fw/version', t.firmwareVersion, {retain:true});
  }
  t.mqttClient.publish(t.mqttTopic + '/$name', t.config.name, {retain:true});
  t.mqttClient.publish(t.mqttTopic + '/$stats', 'interval,uptime', {retain:true});

  if( t.config.mac != null) {
    t.mqttClient.publish(t.mqttTopic + '/$mac', t.config.mac, {retain:true});
  }

  if( t.config.ip != null) {
    t.mqttClient.publish(t.mqttTopic + '/$localip', t.config.ip, {retain:true});  
  }
  
  var nodes = [];
  _.each(t.nodes, function(node){
    var node = node.isRange ? node.name + '[]' : node.name;
    
    nodes.push(node);
  })
  t.mqttClient.publish(t.mqttTopic + '/$nodes', nodes.join(','), {retain:true});
  
  _.each(t.nodes, function(node){
    node.onConnect();
  })
  t.emit('connect');

  // Call the stats interval now, and at regular intervals
  t.onStatsInterval();
  t.interval = setInterval(function() {
    t.onStatsInterval();
  }, t.statsInterval * 1000);

  t.mqttClient.publish(t.mqttTopic + '/$state', 'ready', {retain:true});
}

// Called on mqtt client disconnect
proto.onDisconnect = function() {
  var t = this;
  t.isConnected = false;
  t.interval = clearInterval(t.interval);
  _.each(t.nodes, function(node){
    node.onDisconnect();
  })
  t.emit('disconnect');
}

// Called on every stats interval
proto.onStatsInterval = function() {
  var t = this;
  var uptime = (Date.now() - t.startTime) / 1000;
  t.mqttClient.publish(t.mqttTopic + '/$stats/uptime', '' + _.round(uptime,0), {retain:false});
  t.mqttClient.publish(t.mqttTopic + '/$stats/interval', '' + t.statsInterval, {retain:true});
  _.each(t.nodes, function(node){
    node.onStatsInterval();
  })
  t.emit('stats-interval');
}

// Called when a device message is received
proto.onMessage = function(topic, msg) {
  var t = this;
  var parts = topic.split('/');
  var deviceTopic = parts.slice(2).join('/');

  // Emit broadcast messages to broadcast listeners
  if (parts[1] == '$broadcast') {
    t.emit('broadcast', deviceTopic, msg);
    return;
  }

  // Emit to listeners of all device topics
  t.emit('message', deviceTopic, msg);

  // Emit to listeners of the specific device topic
  t.emit('message:' + deviceTopic, msg);

  // Invoke property setters if this is a property set message
  if (parts[1] == t.name && parts[4] == 'set') {
    var nodeName = parts[2];
    var propName = parts[3];
    var value = msg;
    var range = { isRange: false, index: 0 }

    if(nodeName.indexOf("_") > -1) {
      range.isRange = true;
      var nodeParts = nodeName.split('_');
      nodeName = nodeParts[0];
      range.index = parseInt(nodeParts[1])
    }
    var node = t.nodes[nodeName];

    if (node) {
      if (node.isRange !== range.isRange) {
        return;
      }
      
      var prop = node.props[propName];
      if (prop && _.isFunction(prop.setter)) {
        // This interface is consistent with esp8266 homie
        prop.setter(range, value);
      }
    }
  }
}
